sdks/python/apache_beam/io/gcp/bigquery.py

```diff
@@ -1515,11 +1526,41 @@ def process(self, element, *schema_side_inputs):
 
     if not self.with_batched_input:
       row_and_insert_id = element[1]
+      row_byte_size = get_deep_size(row_and_insert_id)
+
+      # send large rows that exceed BigQuery insert limits to DLQ
+      if row_byte_size >= self.max_insert_payload_size:
+        row_mb_size = row_byte_size / 1_000_000
+        max_mb_size = self.max_insert_payload_size / 1_000_000
+        error = (
+            f"Received row with size {row_mb_size}MB that exceeds "
+            f"the maximum insert payload size set ({max_mb_size}MB).")
+        return [
+            pvalue.TaggedOutput(
+                BigQueryWriteFn.FAILED_ROWS_WITH_ERRORS,
+                GlobalWindows.windowed_value(
+                    (destination, row_and_insert_id[0], error))),
+            pvalue.TaggedOutput(
+                BigQueryWriteFn.FAILED_ROWS,
+                GlobalWindows.windowed_value(
+                    (destination, row_and_insert_id[0])))
+        ]
+
+      buffer_byte_size_with_row = (
+          row_byte_size if not self._rows_buffer[destination] else
+          row_byte_size + get_deep_size(self._rows_buffer[destination]))
```

Comment:
Can't this be expensive?

We already had `get_deep_size` for every element before, perhaps we could also accumulate it here instead of doing `get_deep_size` on the entire array again?

Reply:
In case it wasn't clear, I'm talking about keeping a supporting list like

self._rows_buffer_bytes[destination]